### PR TITLE
Add a builder for route messages

### PR DESCRIPTION
**Pull request by little-dude**
For both the `RouteAddRequest` and `RouteDelRequest`, we need to build
a route message. However only the former provides helper methods to do
so. Instead of duplicating code, create a separate builder, and use it for
both request types.

This MR also allows building route requests parametrized by `IpAddr`, in
addition to the existing `Ipv4Addr` and `Ipv6Addr` implementations.

----

This is a non-backward compatible change.

### CODE REVIEW

File: src/route/builder.rs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        self.message
+            .attributes
+            .push(RouteAttribute::Priority(priority));
+        self
+    }
+
+    /// Sets the route table.
+    ///
+    /// Default is main route table.
+    #[deprecated(note = "Please use `table_id` instead")]
```
Comment (cathay4t) [inline src/route/builder.rs:57]: Since we are breaking the API by replace it with new. There is no need to keep old deprecated function in new struct.

**Review by cathay4t — CHANGES_REQUESTED** [review]
Please also fix `examples/add_route.rs`

**Comment by cathay4t** [comment]
In general, I like the idea of `XxxBuilder` instead old on-line changes of `message_mut()`.

**Comment by richardstephens** [comment]
Hi, I'm jumping into the discussion here on request of @cathay4t in response to my PR number #53 .

Our product uses rtnetlink quite extensively to configure the host's networking. In principle, I would much rather use Builder's that make the intent of the netlink command we issue clear from the code. However I've sturggled a bit with understanding how to achieve my desired results by reading the netlink documentation. My dev loop often involved `strace`'ing `ip` or `bridge` commands to find out what the netlink message I need looks like.

I'm alsp specifically concerned about the message header flags. In my PR, I was not able to achieve the desired result with a set link message - I needed to issue an 'add' link message with the correct flags even though I was updating an existing object. I'm not sure how to design a clean API for this other than by exposing API to manipulate the flags directly.

I'm happy to make another attempt at that PR in line with the builder approach, along with type-safe setters for the flags. What do you think?

**Comment by cathay4t** [comment]
> Hi, I'm jumping into the discussion here on request of @cathay4t in response to my PR number #53 .
> 
> Our product uses rtnetlink quite extensively to configure the host's networking. In principle, I would much rather use Builder's that make the intent of the netlink command we issue clear from the code. However I've sturggled a bit with understanding how to achieve my desired results by reading the netlink documentation. My dev loop often involved `strace`'ing `ip` or `bridge` commands to find out what the netlink message I need looks like.

There is no good netlink document except kernel code. I also use `strace -s 10000` for ip command initially also, but will cross reference with kernel code.

> 
> I'm also specifically concerned about the message header flags. In my PR, I was not able to achieve the desired result with a set link message - I needed to issue an 'add' link message with the correct flags even though I was updating an existing object. I'm not sure how to design a clean API for this other than by exposing API to manipulate the flags directly.

I will tag new release to `netlink-packit-route` this week which will contain bit changes(sorry) to flags.
Exposing a API at rtnetlink level for flags is simpler for this builder workflow now.

> 
> I'm happy to make another attempt at that PR in line with the builder approach, along with type-safe setters for the flags. What do you think?

Sure, but please wait `netlink-packit-route`  0.20.0 release.

If @little-dude does not have time to move this PR forward, I will append my patches to this PR to get it merged next week.

**Comment by little-dude** [comment]
I'll try to fix this up today @cathay4t. Sorry, it's still busy at work, and I'm working on netlink by spiked.

File: src/route/builder.rs (hunk 1 above)
Comment (little-dude) [reply src/route/builder.rs]: Removed, thanks!

**Comment by little-dude** [comment]
Done, @cathay4t

**Comment by little-dude** [comment]
Hello @cathay4t, would it be possible to get this merged within the next couple weeks by any chance?

**Comment by cathay4t** [comment]
Please give me 2 days to review it.

**Review by cathay4t — APPROVED** [review]

**Comment by cathay4t** [comment]
LGTM. Merged. Thanks for the efforts!

**Comment by little-dude** [comment]
Thanks @cathay4t !

**Comment by little-dude** [comment]
@cathay4t  would it be possible to make a release with these changes? I haven't done a release in a while, and with all the inter-dependencies between the crates I'm a bit scared of making one.

**Comment by little-dude** [comment]
@cathay4t would you be able to make a release? If not I'll give it a try.

**Comment by cathay4t** [comment]
Can we use this builder to other component(addr, link, rule, rc)also before release?

And I am also considering to `RouteGetRequest::new(handle: Handle, rt_msg: RouteMessage)` also, to be consistent.

**Comment by cathay4t** [comment]
I will create a PR tomorrow for expanding building to crate wide. After that, I can make a release.

Meanwhile, please also vote on #66

**Comment by little-dude** [comment]
> After that, I can make a release.

I just realized I'd also need a netlink-packet-route release to get https://github.com/rust-netlink/netlink-packet-route/pull/98 if you don't mind

**Comment by cathay4t** [comment]
netlink-packet-route 0.20.1 released.

Working on expanding the use of builder.

**Comment by little-dude** [comment]
Thanks a ton @cathay4t !

**Comment by cathay4t** [comment]
I am currently work on creating builder for link:

https://github.com/cathay4t/rtnetlink/tree/link_builder

**Pull request merged**